model/labels/regexp.go

```diff
@@ -923,34 +968,55 @@ func optimizeEqualStringMatchers(input StringMatcher, threshold int) StringMatch
 		return true
 	}
 
-	if !findEqualStringMatchers(input, analyseCallback) {
+	analysePrefixMatcherCallback := func(prefix string, prefixCaseSensitive bool, matcher StringMatcher) bool {
+		// Ensure we don't have mixed case sensitivity.
+		if caseSensitiveSet && caseSensitive != prefixCaseSensitive {
+			return false
+		} else if !caseSensitiveSet {
+			caseSensitive = prefixCaseSensitive
+			caseSensitiveSet = true
+		}
+		if len(prefix) < minPrefixLength {
+			minPrefixLength = len(prefix)
+		}
+
+		numPrefixes++
+		return true
+	}
+
+	if !findEqualOrPrefixStringMatchers(input, analyseEqualMatcherCallback, analysePrefixMatcherCallback) {
 		return input
 	}
 
 	// If the number of values found is less than the threshold, then we should skip the optimization.
-	if numValues < threshold {
+	if (numValues + numPrefixes) < threshold {
 		return input
 	}
 
 	// Parse again the input StringMatcher to extract all values and storing them.
 	// We can skip the case sensitivity check because we've already checked it and
 	// if the code reach this point then it means all matchers have the same case sensitivity.
-	multiMatcher := newEqualMultiStringMatcher(caseSensitive, numValues)
+	multiMatcher := newEqualMultiStringMatcher(caseSensitive, numValues, numPrefixes, minPrefixLength)
```

Comment:
If there's no prefix we pass `minPrefixLength=math.MaxInt`. Then, `equalMultiStringMapMatcher.Matches()` will check for `if m.minPrefixLen > 0`, which is true even if we're in the case "there no prefix". Likewise, the safety check done by `equalMultiStringMapMatcher.addPrefix()` is skrew up by this.

One way to simply fix this is:
- Initialise `minPrefixLength` to `0`
- Change `if len(prefix) < minPrefixLength {` to `if numPrefixes == 0 || (len(prefix) < minPrefixLength) {`